stable-diffusion/main.py

```diff
@@ -0,0 +1,167 @@
+        tokenizer=model.tokenizer,
+        mean=config.dataset.mean,
+        std=config.dataset.std,
+        image_column=config.dataset.image_column,
+        caption_column=config.dataset.caption_column,
+        batch_size=train_batch_size)
+
+    # Eval dataset
+    eval_dataspec = build_prompt_dataspec(config.dataset.prompts,
+                                          batch_size=eval_batch_size)
+
+    # Optimizer
+    print('Building optimizer and learning rate scheduler')
+
+    optimizer = torch.optim.AdamW(params=model.parameters(),
+                                  lr=config.optimizer.lr,
+                                  weight_decay=config.optimizer.weight_decay)
+
+    # Learning rate scheduler: LR warmup for 8 epochs, then cosine decay for the rest of training
+    lr_scheduler = ConstantScheduler()
+    print('Built optimizer and learning rate scheduler\n')
+
+    
+    loggers = [
+        build_logger(name, logger_config)
+        for name, logger_config in config.loggers.items()
+    ]
+    # Callbacks for logging
+    print('Building Speed, LR, and Memory monitoring callbacks')
+    speed_monitor = SpeedMonitor(
+        window_size=50
+    )  # Measures throughput as samples/sec and tracks total training time
+    lr_monitor = LRMonitor()  # Logs the learning rate
+    memory_monitor = MemoryMonitor()  # Logs memory utilization
+
+    image_logger = LogDiffusionImages()  # Logs images generated from prompts in the eval set
+
+    # Callback for checkpointing
+    print('Built Speed, LR, and Memory monitoring callbacks\n')
+
+    print('Building algorithms')
+    if config.use_ema:
+        algorithms = [EMA(half_life='100ba', update_interval='20ba')]
+    else:
+        algorithms = None
+    print('Built algorithms')
+
+    # Create the Trainer!
+    print('Building Trainer')
+    device = 'gpu' if torch.cuda.is_available() else 'cpu'
+    precision = 'amp' if device == 'gpu' else 'fp32'  # Mixed precision for fast training when using a GPU
+    trainer = Trainer(
+        run_name=config.run_name,
+        model=model,
+        train_dataloader=train_dataspec,
+        eval_dataloader=eval_dataspec,
+        optimizers=optimizer,
+        schedulers=lr_scheduler,
+        algorithms=algorithms,
+        loggers=loggers,
+        max_duration=config.max_duration,
+        eval_interval=config.eval_interval,
+        callbacks=[speed_monitor, lr_monitor, memory_monitor, image_logger],
+        save_folder=config.save_folder,
+        save_interval=config.save_interval,
+        save_num_checkpoints_to_keep=config.save_num_checkpoints_to_keep,
+        load_path=config.load_path,
+        device=device,
+        precision=precision,
+        grad_accum=config.grad_accum,
+        seed=config.seed)
+    print('Built Trainer\n')
+
+    print('Logging config')
+    log_config(config)
+
+    print('Train!')
+    trainer.fit()
+    # Return trainer for testing purposes
+    return trainer
```

Comment:
super minor; just trying to shave off a line of code
```suggestion
    return trainer  # Return trainer for testing purposes
```